lib/SPIRV/SPIRVReader.cpp

```diff
@@ -4364,6 +4364,51 @@ bool llvm::getSpecConstInfo(std::istream &IS,
   return !IS.bad();
 }
 
+bool llvm::readSPIRVHeader(std::istream &IS, SPIRVHeaderData &Result,
+                           std::string &ErrMsg) {
+  std::unique_ptr<SPIRVModule> BM(SPIRVModule::createSPIRVModule());
+  BM->setAutoAddExtensions(false);
+  SPIRVDecoder D(IS, *BM);
+  SPIRVWord Magic;
+  SPIRVWord Buffer;
+  bool Success;
+  D >> Magic;
+  if (!BM->getErrorLog().checkError(Magic == MagicNumber, SPIRVEC_InvalidModule,
+                                    "invalid magic number")) {
+    ErrMsg = "invalid magic number";
+    return false;
+  }
+
+  D >> Buffer;
+  Result.Version = Buffer;
+  D.ignore(3);
+  while (D.getWordCountAndOpCode()) {
+    Success = 0;
+    switch (D.OpCode) {
+    case spv::OpMemoryModel:
+      D.ignore(1);
+      Success = D.getNextSPIRVWord(Buffer);
+      if (Success)
+        Result.MemoryModel = Buffer;
+      break;
```

Comment:
Can't we just return here, once we have seen `OpMemoryModel`?  See "2.4. Logical Layout of a Module" of the SPIR-V spec.